sourcesapi/resulttypes.py

```diff
@@ -0,0 +1,185 @@
+from __future__ import annotations
```

Comment:
We should have a generic set that the sources API will use which will end up being a superset of what any individual source will provide and does not need to be structured like CV or any other api.

I initially created the resulttypes.py file purely to stop recursive imports so it never really had a designed purpose it was just somewhere to store some classes

Reply:
My thinking is a search results one and an issues result one. The full info needs to be formed to the metadata standard obviously.
```
class ComicSearchResult(TypedDict):
    count_of_issues: NotRequired[int]
    description: NotRequired[str]
    id: int
    image: NotRequired[str]
    name: str
    publisher: NotRequired[ComicPublisher]
    start_year: NotRequired[str]
```
It's probably not possible to get away with all the not required currently. 
```
class ComicFetchIssuesByVolumeResults(TypedDict):
    cover_date: str
    description: str
    id: int
    image: str
    issue_number: str
    name: str
    site_detail_url: str
```
Those two seem to be all that's required?

Reply:
The NotRequired and even the whole point of using the TypedDict is that I started using [mypy] (it helps find plenty of errors before I commit them) but it's more like a best practice and I don't have the CI check it so it's not a blocker.
also we can change it to:
```python
class ComicSearchResult(TypedDict, total=False):
    count_of_issues: int
    description: str
    id: Required[int]
    image: str
    name: Required[str]
    publisher: ComicPublisher
    start_year: str
```

Besides that we have two things to consider:
1. Cache
  - Do we do the caching CT side or Talker side?
    Are these the results that we are expecting the talker to give us and we just cache them every time? or do we expect the Talker to do the caching because it would know more about its source?
1. GenericMetadata
  - Do we call a function on the class that returns the metadata? or do we have a single function that combines the metadata from these definitions?

I would pick an answer to those questions and then try implementing it, we won't know which is better until we try one of them.

[mypy]: https://mypy-lang.org

Reply:
I've been looking at this and this is my thinking etc.:

I've flattened "publisher" to a str. Looking in the code, I could not find an instance where anything other than the name is used.

I think letting the talker handle the cache is the way to go but it still has to conform to the ComicSearchResults etc. data format. The reason is that the talker may want to overload a field to save JSON info etc. This example might not work but aliases could be stored. You could add an option in the talker to prefer the alias name closest to the search term or a specific language etc.

As for GenericMetadata. In the base class there is map_data_to_metadata which should output GenericMetadata. The API has fetch_issue_data which expects GenericMetadata out. The map_data_to_metadata in the base class is more a suggestion than requirement and could be removed. The important part is that fetch_issue_data returns GenericMetadata.

Just a little heads up before the push.